### DISCUSSION TITLE
Best way to dynamically load hundreds of components?

### DISCUSSION THREAD

**Discussion by FBosler** [vercel/next.js#13837] [Help]
Let's assume that you have a very dynamic app with hundreds (if not thousands) of components, the components are sufficiently different such that they can not be refactored into fewer components. (say for example different charts, and interactions and layouts).

Let's further assume that the component to be displayed depends on a user-input (i.e. can not be known at build time).

I am wondering what the most performant/stable way would be for a user to have access to all components without killing the overall experience (in terms of loading times). I don't care so much about an increase in build time.

**Disclaimer:** There has been another [discussion](https://github.com/vercel/next.js/discussions/11291?sort_order=chronological) around the topic. However, what's discussed there is imho only feasible for a limited number of components. 

It was stated there that:
> Using `import(`../../components/${d}`)` causes every single file under that directory to be bundled as a separate file. In 99% of the cases, this is not what you want as it causes significant performance issues with regards to builds.

I am not quite sure what exactly that means in terms of user experience. Why would it be bad if every component would be its own file? Isn't that what code-splitting is about? Load only what you need when you need it?

The suggested solution in the other thread is to build a component map and then look them up:

```javascript
const componentList = {
  MyComponent: dynamic(() => import(`../../components/my-component`)),
  OtherComponent: dynamic(() => import(`../../components/other-component`)),
  or more generally:
  COMPONENTKEY dynamic(() => import(COMPONENTPATH))
}
```

The problem that I am having with this approach is that this list would grow extremely long and become very hard to maintain. I'd much rather use a database (where I have meta-information on the components, which also includes COMPONENTKEY and COMPONENTPATH.)

**Comment by FBosler** [comment]
Any takers on this? I'd love to discuss pro's and con's of different approaches.

**Comment by TommySorensen** [comment]
This should enable bundle splitting by using the dynamic from Next.js. Let me know if it works.

```js
// In your page call a component:
const modules = [
  { kind: 'module1', data: 'somecontent'}
  { kind: 'module2', data: 'somecontent'}
  { kind: 'module3', data: 'somecontent'}
]
<DynamicModules modules={modules} />

// DynamicModules.js component:
import dynamic from 'next/dynamic'

const moduleList = {
  'module1': dynamic(() => import('/components/module1')),
  'module2': dynamic(() => import('/components/module2')),
  'module3': dynamic(() => import('/components/module3')),
}

const DynamicModules = ({ modules }) => {
  // Creating a list of components causes webpack only to import the components that are really needed
  const components = modules?.map(module => moduleList[module.kind])

  return (
    {components?.map((Component, i) => (
      <Component
        key={i}
        {...modules[i]}
      />
    ))}
  )
}
```

**Reply by FBosler** [reply]
@ThommySorenson Really appreciate the reply. 

I think I might not have been 100% clear on my problem. The way I am currently doing it, already gives bundle splitting.

The problem I am having is that this list keeps growing (and will eventually be thousands of entries - which makes management extremely hard). 
I'd really love to generate this ComponentList as part of the build process via means of extracting the `COMPONENTKEY` and `COMPONENTPATH` from the database.

```
const componentList = {
  MyComponent: dynamic(() => import(`../../components/my-component`)),
  OtherComponent: dynamic(() => import(`../../components/other-component`)),
  or more generally:
  COMPONENTKEY dynamic(() => import(COMPONENTPATH))
}
```

**Additional Info:** I really have to build all components (because I will not know at build time, which components are going to be used at runtime)

**Reply by TommySorensen** [reply]
I'm not sure what you want. You mention that you don't know at build time what components to be used. So is the components based on the user? If so, then just use the normal way of using `dynamic()` where you just lazy load the components for that user, right?

**Reply by TommySorensen** [reply]
Something like [this](https://nextjs.org/docs/advanced-features/dynamic-import):
```js
import dynamic from 'next/dynamic'

const DynamicComponentWithNoSSR = dynamic(
  () => import('../components/hello3'),
  { ssr: false }
)

function Home() {
  return (
    <div>
      <Header />
      <DynamicComponentWithNoSSR />
      <p>HOME PAGE is here!</p>
    </div>
  )
}
```

**Reply by FBosler** [reply]
Hey Thommy,

yea, I think it’s a bit of a special case. In essence I have thousands of components that should be available to users at run time (based on their inputs)

I also want to make sure that only acceptable components can be selected, which is why I Need some reference list (as indicated in my initial example).

Ideally - to make this process resilient - I would build that reference list at build time.

In your example it seems like I already have to know the component to import at build time, no?

**Reply by TommySorensen** [reply]
I guess you can fetch the list of components in your `/pages/home.js` file in the `getStaticProps` or `getServerSideProps` and then pass it as props to you page like:
```js
// /pages/home.js

export default function Home({ componentList }) {
  // Create the list of components based of componentList here. Here i just took the first one
  const moduleList = {
    componentList[0].name: dynamic(() => import(componentList[0].path))
  }

  const components = modules?.map(module => moduleList[module.kind])

  return (
    {components?.map((Component, i) => (
      <Component
        key={i}
        {...modules[i]}
      />
    ))}
  )
}


export async function getStaticProps(context) {
  return {
    props: {
      componentList: await getComponentList(context),
    }
  }
}
```
I have no idea if it works, but i guess it should. Dont know if is working with bundle splitting. Your `async getComponentList` function will then know the list of components from somewhere else, like a cms? Try it out 😃

Update: just realised you want this based on user inputs, so this seems like the wrong way to do it

**Reply by TommySorensen** [reply]
It sound like you maybe wanna do this in some kind of custom webpack setup. In the first place i don't understand why you have thousands of components. Don't you have page types or something like that? Or is this a news website where every single component is a new way of showing some content? Let me know why so many components is necessary 🙂

**Reply by FBosler** [reply]
Will give this a try and let you know how it goes!! Yea in essence it’s an app that will display the different types of data quite differently. With a high degree of abstraction I might be able to bring it down to hundreds of components. Although that would require substantial rewrites and redesign.

The idea is that it’s sort of a chat, where different information (Think Math concepts) will be displayed differently.

Thanks for your feedback! Really appreciate it.

**Reply by moarief** [reply]
> This should enable bundle splitting by using the dynamic from Next.js. Let me know if it works.
> 
> ```js
> // In your page call a component:
> const modules = [
>   { kind: 'module1', data: 'somecontent'}
>   { kind: 'module2', data: 'somecontent'}
>   { kind: 'module3', data: 'somecontent'}
> ]
> <DynamicModules modules={modules} />
> 
> // DynamicModules.js component:
> import dynamic from 'next/dynamic'
> 
> const moduleList = {
>   'module1': dynamic(() => import('/components/module1')),
>   'module2': dynamic(() => import('/components/module2')),
>   'module3': dynamic(() => import('/components/module3')),
> }
> 
> const DynamicModules = ({ modules }) => {
>   // Creating a list of components causes webpack only to import the components that are really needed
>   const components = modules?.map(module => moduleList[module.kind])
> 
>   return (
>     {components?.map((Component, i) => (
>       <Component
>         key={i}
>         {...modules[i]}
>       />
>     ))}
>   )
> }
> ```

How to convert this to Typescript?

**Reply by moarief** [reply]
Was simple :)

```
type DynamicModulesProps = {
  modules: { kind: string; data: string }[];
};

const DynamicModules: React.FC<DynamicModulesProps> = ({ modules }) => {
  // Creating a list of components causes webpack only to import the components that are really needed
  const components: React.ComponentType<{}>[] = modules?.map(
    (module) => moduleList[module.kind]
  );

  return (
    <>
      {components?.map((Component, i) => (
        <Component key={i} {...modules[i]} />
      ))}
    </>
  );
};

export default DynamicModules;
```
